llvm/lib/Target/SPIRV/SPIRVEmitNonSemanticDI.cpp

```diff
@@ -156,22 +205,56 @@ bool SPIRVEmitNonSemanticDI::emitGlobalDI(MachineFunction &MF) {
         GR->buildConstantInt(SourceLanguage, MIRBuilder, I32Ty, false);
 
-    // Emit DebugCompilationUnit
+    [[maybe_unused]]
     const Register DebugCompUnitResIdReg =
-        MRI.createVirtualRegister(&SPIRV::IDRegClass);
-    MRI.setType(DebugCompUnitResIdReg, LLT::scalar(32));
-    MIB = MIRBuilder.buildInstr(SPIRV::OpExtInst)
-              .addDef(DebugCompUnitResIdReg)
-              .addUse(GR->getSPIRVTypeID(VoidTy))
-              .addImm(static_cast<int64_t>(
-                  SPIRV::InstructionSet::NonSemantic_Shader_DebugInfo_100))
-              .addImm(SPIRV::NonSemanticExtInst::DebugCompilationUnit)
-              .addUse(DebugInfoVersionReg)
-              .addUse(DwarfVersionReg)
-              .addUse(DebugSourceResIdReg)
-              .addUse(SourceLanguageReg);
-    MIB.constrainAllUses(*TII, *TRI, *RBI);
-    GR->assignSPIRVTypeToVReg(VoidTy, DebugCompUnitResIdReg, MF);
+        EmitDIInstruction(SPIRV::NonSemanticExtInst::DebugCompilationUnit,
+                          {DebugInfoVersionReg, DwarfVersionReg,
+                           DebugSourceResIdReg, SourceLanguageReg});
+
+    // Emit DebugInfoNone. This instruction is a wildcard accepted
+    // by standard to put into instruction arguments as Not Available/Null
+    const Register DebugInfoNoneReg =
+        EmitDIInstruction(SPIRV::NonSemanticExtInst::DebugInfoNone, {});
+
+    for (auto *BasicType : BasicTypes) {
+      const Register BasicTypeStrReg = EmitOpString(BasicType->getName());
+
+      const Register ConstIntBitwidthReg = GR->buildConstantInt(
+          BasicType->getSizeInBits(), MIRBuilder, I32Ty, false);
+
+      uint64_t AttributeEncoding = BaseTypeAttributeEncoding::Unspecified;
+      switch (BasicType->getEncoding()) {
+      case dwarf::DW_ATE_signed:
+        AttributeEncoding = BaseTypeAttributeEncoding::Signed;
+        break;
+      case dwarf::DW_ATE_unsigned:
+        AttributeEncoding = BaseTypeAttributeEncoding::Unsigned;
+        break;
+      case dwarf::DW_ATE_unsigned_char:
+        AttributeEncoding = BaseTypeAttributeEncoding::UnsignedChar;
+        break;
+      case dwarf::DW_ATE_signed_char:
+        AttributeEncoding = BaseTypeAttributeEncoding::SignedChar;
+        break;
+      case dwarf::DW_ATE_float:
+        AttributeEncoding = BaseTypeAttributeEncoding::Float;
+        break;
+      case dwarf::DW_ATE_boolean:
+        AttributeEncoding = BaseTypeAttributeEncoding::Boolean;
+        break;
+      case dwarf::DW_ATE_address:
+        AttributeEncoding = BaseTypeAttributeEncoding::Address;
+      }
+
+      const Register AttributeEncodingReg =
+          GR->buildConstantInt(AttributeEncoding, MIRBuilder, I32Ty, false);
+
+      [[maybe_unused]]
+      const Register BasicTypeReg =
+          EmitDIInstruction(SPIRV::NonSemanticExtInst::DebugTypeBasic,
+                            {BasicTypeStrReg, ConstIntBitwidthReg,
+                             AttributeEncodingReg, DebugInfoNoneReg});
```

Comment:
I see in https://github.com/KhronosGroup/SPIRV-Registry/blob/main/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc#DebugTypeBasic that the `<id> Flags` operand is required to be "... the <id> of a 32-bit integer OpConstant ...". Can you please clarify why a result of `OpExtInst` is used in this case instead?

Reply:
I don't see how this conforms to the specification. As far as I see, it violates the SPIR-V spec.

Reply:
Per https://github.com/KhronosGroup/SPIRV-Registry/blob/main/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc#missing-debugging-information 
"Other instructions can refer to this one in case the debugging information is unknown, not available, or not applicable." While indeed other instructions doesn't say it explicitly , that a particular operand can also be DebugInfoNone (and not only OpConstant or some other Debug instruction) - it make sense for some debug information to be missing.

Reply:
Great @MrSidims , thank you for the clarification!

Reply:
@MrSidims Please see https://github.com/KhronosGroup/SPIRV-Registry/issues/280#issuecomment-2346209669

Reply:
@VyacheslavLevytskyy Replaced `DebugInfoNone` to zero which also passes spir-val. This effectively ends the PR involvement in the `DebugInfoNone` compliance discussion.

Reply:
@VyacheslavLevytskyy thanks for reviving the github issue!